packages/darklang/stdlib/string.dark

```diff
@@ -167,4 +169,145 @@ module Darklang =
+        | _, "" ->
+          s
+          |> String.toList
+          |> List.map (fun c ->
+            PACKAGE.Darklang.Stdlib.Char.toString c ++ replaceWith)
+          |> String.join ""
+          |> PACKAGE.Darklang.Stdlib.String.prepend replaceWith
+        | _ -> s |> String.split searchFor |> String.join replaceWith
+
+
+      /// Returns the length of the string
+      let length (s: String) : Int =
+        s |> String.toList |> PACKAGE.Darklang.Stdlib.List.length
+
+
+      // is it bad to do it this way?
+      /// Concatenates the two strings by appending <param s2> to <param s1> and returns the joined string.
+      let append (s1: String) (s2: String) : String = $"{s1}{s2}"
+
+
+      /// Reverses <param string>
+      let reverse (string: String) : String =
+        string
+        |> String.toList
+        |> PACKAGE.Darklang.Stdlib.List.reverse
+        |> PACKAGE.Darklang.Stdlib.String.fromList
+
+
+      /// Combines a list of strings with the provided separator
+      let join (l: List<String>) (separator: String) : String =
+        (List.map l (fun s -> s ++ separator)) |> String.join ""
+
+
+      /// Escape an untrusted string in order to include it safely in HTML output
+      let htmlEscape (html: String) : String =
+        html
+        |> String.toList
+        |> List.map (fun c ->
+          match c with
+          | '<' -> "&lt;"
+          | '>' -> "&gt;"
+          | '&' -> "&amp;"
+          // include these for html-attribute-escaping
+          // even though they're not strictly necessary
+          // for html-escaping proper.
+          | '"' -> "&quot;"
+          // &apos; doesn't work in IE....
+          | ''' -> "&#x27;"
+          | _ -> PACKAGE.Darklang.Stdlib.Char.toString c)
+        |> PACKAGE.Darklang.Stdlib.String.join ""
+
+
+      // this is a helper function for slice
+      let normalize (i: Int) (len: Int) : Int =
+        (if i < 0 then len + i else i)
+        |> PACKAGE.Darklang.Stdlib.Int.min len
+        |> PACKAGE.Darklang.Stdlib.Int.max 0
+
+
+      /// Returns the substring of <param string> between the <param from> and <param to> indices.
+      /// Negative indices start counting from the end of <param string>.
+      let slice (s: String) (f: Int) (l: Int) : String =
+        let len = PACKAGE.Darklang.Stdlib.String.length s
+        let f = f |> PACKAGE.Darklang.Stdlib.String.normalize len
+        let l = l |> PACKAGE.Darklang.Stdlib.String.normalize len
+        let l = if f > l then f else l
+
+        s
+        |> String.toList
+        |> PACKAGE.Darklang.Stdlib.List.drop f
+        |> PACKAGE.Darklang.Stdlib.List.take (l - f)
+        |> PACKAGE.Darklang.Stdlib.String.fromList
+
+
+      /// Returns a copy of <param str> with all leading whitespace removed. 'whitespace'
+      /// here means all Unicode characters with the {{White_Space}} property, which
+      /// includes {{\" \"}}, {{\"\\t\"}} and {{\"\\n\"}}
+      let trimStart (str: String) : String =
```

Comment:
I'm not sure the logic is right - there's lots of Unicode whitespace characters, right? Should probably Char.isWhitespace or something.